burn-core/src/nn/initializer.rs

```diff
@@ -10,74 +10,116 @@ use crate as burn;
 /// Enum specifying with what values a tensor should be initialized
 #[derive(Config, Debug, PartialEq)]
 pub enum Initializer {
-    /// Fills tensor with values drawn uniformly between specified values
-    Uniform(f64, f64),
-    /// Must be implemented by caller. TODO change to NormalizedUniform, fills tensor with values drawn uniformly between -sqrt(1/fan_in) and sqrt(1/fan_in).
-    UniformDefault,
-    /// Fills tensor with values drawn from normal distribution with specified mean and std
-    Normal(f64, f64),
     /// Fills tensor with specified value everywhere
     Constant(f64),
     /// Fills tensor with 1s everywhere
     Ones,
     /// Fills tensor with 0s everywhere
     Zeros,
+    /// Fills tensor with values drawn uniformly between specified values
+    Uniform(f64, f64),
+    /// Fills tensor with values drawn uniformly between -sqrt(1/fan_in) and sqrt(1/fan_in).
+    NormalizedUniform,
```

Comment:
Thanks for the refactor! I was unsure about the origin of this uniform distribution during my own refactoring. However, I believe I now understand it. I've been trying to trace the origins of this initialization technique, and it appears to be based on Kaiming uniform initialization for linear layers, not XavierUniform as Nathaniel initially thought:

https://discuss.pytorch.org/t/clarity-on-default-initialization-in-pytorch/84696

> 1. In the Kaiming uniform initialization, `negative_slope=sqrt(5)` so the `gain=sqrt(2/6)=1/sqrt(3)`. If we substitute this into the `bound` formula, we get `bound = [1/sqrt(3)] * [sqrt(3/ fan_in)]` which simplifies to `bound = 1/sqrt(fan_in)`, which can be represented as `bound^2 = 1 / fan_in`.
> 2. In the linear implementation, `k= 1/in_features`, which in the case of Kaiming can be represented as `k=1/fan_in`. We want a boundary of `[-sqrt(k), sqrt(k)]` where `k = bound^2 = 1 / fan_in` from step 1.

Given this information, it may be worthwhile to introduce a proper Kaiming initialization and replace our current method:

1. Uniform: https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.kaiming_uniform_
2. Normal: https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.kaiming_normal_

I understand that this might significantly disrupt your current PR, so feel free to address this in a separate PR or file an issue ticket to handle it later. I'll leave that decision to you.

Reply:
@nathanielsimard Let us know of your thoughts as well.